documents/sim-objects/branch-predictors.md

```diff
@@ -0,0 +1,76 @@
+
+### LTAGE Predictor
+TAGE (TAgged GEometric)is a global-history based branch predictor. It features a PC-indexed bimodal predictor and N partially tagged tables, indexed with a hash of the PC and the global branch history. The different lengths of global branch history used to index the partially tagged tables grow geometrically. A small path history is also used in the hash. L-TAGE also features a loop predictor that records iteration count of loops and predicts accordingly.
+
+All TAGE tables are accessed in parallel, and the one using the longest history that matches provides the prediction (some exceptions apply).Entries are allocated in components using a longer history than the one that predicted when the prediction is incorrect.
+
+## Benchmark Characterization
+Control flow benchmarks from Vertical Research Group's [micro-benchmark suite](https://github.com/darchr/microbench) were used to study the branch predictors 
+
+CCa - completely baised branches\
+CCm - heavly baised branches\
+CCe - easy to predict -- branch pattern (10101010...)\
+CCh - impossible to predict -- random branch pattern\
+CCh_st - impossible to predict with a store operation\
+CCl - impossible to predict with large penality -- large airthmetic work per branch\
+CS1 - switch Case Statement of size 10 -- Different Case each time\
+CS3 - switch Case Statement of Size 10 -- Different case every third time\
+
+
+
+### Branch Missprediction 
+![Branch Miss Rate](images/branchMissRate_timingSimple.png)
+
+- All BPs perform as expected in the case of easy to predict benchmarks such as `CCa`, `CCm` and `CCe`, with an exception only of local BP. This is because, local BP uses 2bit counters in its history table for direction prediction, pattern such as 10101010.. in terms branch behaviour is a limiting sequence for a two bit counter. But we see just 10% miss rate because, the for loops in the benchmark are easy predictable. Same reason also applies to immpossible to predict benchmarks such as `CCh`, `CCh_st` and `CCl`.
+```c
+int loop(int zero) {
+  int t = 0,i,iter;
+  for(iter=0; iter < ITERS; ++iter) {
+    for(i=0; i < STEP + zero; i+=1) {
+      if(randArr[i])  {
+        t+=3+3*t;
+      } else {
+        t-=1-5*t;
+      }
+    }
+  }
+  return t;
+}
+```
+- High miss prediction rate in case of control recursive benchmarks such as `CF1`, `CRd`, `CRf` and `CRm` is due the small RAS size(16) used during the experiment.
+- Among the four BPs studied LTAGE gives the best overall performance and Tournament gives the best performance in case control recursive benchmarks.
+![Branch Miss Rate](images/branchMissRate_heatmap_comp.png)
+
+### Impact of target buffer
+gem5 by default uses indirect predictor for branch target prediction of conditional branches, the below figure shows the impact of target buffer choice on control flow benchmarks. BTB seems to work better than indirect predictor because these benchmarks do not have enough indirect branches. 
```

Comment:
Little bit nit-picky here: what do you mean by "do not have enough indirect branches", does that mean the length of the indirect branches or the number of indirect branches?